### ISSUE TITLE
Inconsistencies in generated module XML across different platforms

### ISSUE THREAD

**Issue by jamesnetherton** [smartics/smartics-jboss-modules-maven-plugin#15]
The order of processing of each Smartics module definition is not consistent across different platforms. For example if I configure the plugin to work against a directory structure like:
```
etc/smartics/a.xml
etc/smartics/b.xml
etc/smartics/c.xml
```
The order of processing for these files will be different, depending on the platform being run on (E.g On some OS's a file list() operation will return in alphabetical order but will not on others). 

This difference in process order can impact the results of the generated module.xml files. i.e different platforms will end up with differences in generated files.

**Comment by tdiesler** [comment]
What is the proposed solution here?

**Comment by jamesnetherton** [comment]
Make sure the modules are processed in the same order regardless of the platform they're running on. E.g by sorting the modules collection in [initModules()](https://github.com/smartics/smartics-jboss-modules-maven-plugin/blob/master/src/main/java/de/smartics/maven/plugin/jboss/modules/JBossModulesArchiveMojo.java#L389).

The problem is actually twofold. There's the order of processing mentioned previously and in addition there's a problem with how things are handled in [addSortedDependencies](https://github.com/smartics/smartics-jboss-modules-maven-plugin/blob/f2c7b3206ca8d59da40a6b530a04045c3bb1365b/src/main/java/de/smartics/maven/plugin/jboss/modules/xml/ModuleXmlBuilder.java#L438-L460). 

`sorted.add()` is invoked without any checks, which can cause module dependency information to be changed and overwritten. This introduces additional inconsistencies between platforms.

My original solution to this is [here](https://github.com/jamesnetherton/smartics-jboss-modules-maven-plugin/blob/205ec7fd70d122e74a71e766d2327a7c75923315/src/main/java/de/smartics/maven/plugin/jboss/modules/xml/ModuleXmlBuilder.java#L475-L521).

**Comment by jamesnetherton** [comment]
The other issue seems to be with the order that modules are defined within the module descriptors. Consider the [services integration test](https://github.com/smartics/smartics-jboss-modules-maven-plugin/blob/master/src/it/services/src/etc/jboss-modules/modules.xml).

If the order of the module declarations is reversed then the test case fails. Is this to be expected? The documentation doesn't mention anything about the ordering being important.

**Comment by tdiesler** [comment]
Module defs spread over multiple files should produce the same result as if they were defined in one file. With just one file we would have a defined order (i.e. top-to-bottom). With multiple files, this human defined order is lost.

An overall sorting algorithm that orders modules defs before they are processed is likely needed.  As a first shot, we could try something like

* skipped first
* no wildcard defs processed depth first (i.e. deps before parent)
* wildcard defs last
* multiple matching wildcard defs is an error

**Comment by tdiesler** [comment]
An alternative could be to strictly use human defined module ordering, in which case we would need documented rules about file ordering

**Comment by jamesnetherton** [comment]
I spent a little time looking a bit deeper into this. For the majority of use cases, the plugin seems to work fine regardless of the order that modules are processed in. 

One use case that is an issue, I already described earlier in this thread. The order of processing can make a difference to the outcome of the generated files where wildcard expressions are being used. E.g if the module definitions in [this test case](https://github.com/smartics/smartics-jboss-modules-maven-plugin/blob/master/src/it/services/src/etc/jboss-modules/v1/modules.xml) are reversed, then the desired `services="import"` attribute is never applied to the `org.apache.commons.lang` dependency.

The problem lies in [this code block](https://github.com/smartics/smartics-jboss-modules-maven-plugin/blob/master/src/main/java/de/smartics/maven/plugin/jboss/modules/domain/ModuleMap.java#L258-L266) within the `ModuleMap` class. If the `matchContext.hasGroupMatch()` condition evaluates to true, we go ahead and create a module definition using the wildcarded module definition as a template. We should check to see if the we already have a matching module definition for the module name resolved from the wildcard expression.

I believe the other issues I mentioned have already been resolved by #19 and #21.
